test/e2e/clusterctl_upgrade.go

```diff
@@ -575,16 +585,20 @@ func waitForClusterDeletedV1alpha4(ctx context.Context, input waitForClusterDele
 	}, intervals...).Should(BeTrue())
 }
 
-func machinesMatch(oldMachineList *clusterv1alpha3.MachineList, newMachineList *clusterv1.MachineList) bool {
-	if len(oldMachineList.Items) != len(newMachineList.Items) {
+func matchObjects(oldObjects []client.Object, newObjects []client.Object) bool {
+	for len(oldObjects) != len(newObjects) {
 		return false
 	}
 
```

Comment:
Q: WDYT about just building two sets, one with the old one with the new objects and then using the methods of the set to determine if all old objects still exist / there are new objects which aren't in old?